#### Annual Sales Tax Revenue Analysis

### [Data Source](https://tax.utah.gov/econstats/sales)

### Cal Year

### Notes:


In [1]:
import pandas as pd
import os
import numpy as np
import arcpy
from arcgis.features import GeoAccessor, GeoSeriesAccessor
arcpy.env.overwriteOutput = True

In [2]:
# show all columns
pd.options.display.max_columns = None

In [3]:
# Removes misc characters from string
def replace(string):
    string = string.replace("&", "")
    string = string.replace("-", " ")
    string = string.replace("(", "")
    string = string.replace(")", "")
    string = string.replace("/", " ")
    string = string.replace(",", "")
    string = string.replace(".", "")
    string = string.replace(" ", "_")
    return string


### Load Cal Year 2020
2020 sales tax revenue data

In [4]:
# load excel sheet into a dataframe
xlsx = '.\\Inputs\\2020-annual-sales.xlsx'
calyr_2020 = pd.read_excel(xlsx, sheet_name='Table 9', header=5)

# Drop last empty two rows
calyr_2020.drop(calyr_2020.tail(2).index,inplace=True)

# forward fill values from merged cells
calyr_2020['County'] = calyr_2020['County'].fillna(method='ffill')
calyr_2020['Location Code'] = calyr_2020['Location Code'].fillna(method='ffill')
calyr_2020['City'] = calyr_2020['City'].fillna(method='ffill')

# Figures with less than 10 taxpayers have been rounded up per Tax Commission disclosure rules.
# Rename column and convert values to boolean
calyr_2020.rename(columns={'Unnamed: 5':'Rounded Up'}, inplace=True)
crosswalk = {'*':True, np.nan:False} # County_ID
calyr_2020['Rounded Up'] = calyr_2020['Rounded Up'].map(crosswalk)

calyr_2020.head()


County  Location Code         City  \
0  Beaver County         1002.0  Beaver City   
1  Beaver County         1002.0  Beaver City   
2  Beaver County         1002.0  Beaver City   
3  Beaver County         1002.0  Beaver City   
4  Beaver County         1002.0  Beaver City   

                     Economic Sector (NAICS Code)     CY2020  Rounded Up  
0   AGRICULTURE, FORESTRY, FISHING & HUNTING (11)   500000.0        True  
1  MINING, QUARRYING, & OIL & GAS EXTRACTION (21)     3000.0        True  
2                                  UTILITIES (22)  2750000.0        True  
3                               CONSTRUCTION (23)   258778.0       False  
4                           MANUFACTURING (31-33)  6259592.0       False

In [5]:
# pivot the table to get sectors as columns with tax values by city
sectors_calyr_2020 = pd.pivot_table(calyr_2020,values = 'CY2020',index ='City', columns = 'Economic Sector (NAICS Code)', 
                                 aggfunc='first')

# fill NAs with 0
sectors_calyr_2020 = sectors_calyr_2020.fillna(0)

In [6]:
sectors_calyr_2020.columns

Index(['ACCOMMODATION (721)',
       'ADMIN. & SUPPORT & WASTE MANAG. & REMED. SERVICES (56)',
       'AGRICULTURE, FORESTRY, FISHING & HUNTING (11)',
       'ARTS, ENTERTAINMENT AND RECREATION (71)', 'CONSTRUCTION (23)',
       'EDUCATIONAL SERVICES (61)', 'FINANCE & INSURANCE (52)',
       'FOOD SERVICES & DRINKING PLACES (722)',
       'HEALTH CARE & SOCIAL ASSISTANCE (62)', 'INFORMATION (51)',
       'MANAGEMENT OF COMPANIES & ENTERPRISES (55)', 'MANUFACTURING (31-33)',
       'MINING, QUARRYING, & OIL & GAS EXTRACTION (21)',
       'OTHER SERVICES-EXCEPT PUBLIC ADMINISTRATION (81)',
       'PRIOR-PERIOD PAYMENTS & REFUNDS', 'PRIVATE MOTOR VEHICLE SALES',
       'PROFESSIONAL, SCIENTIFIC & TECHNICAL SERVICES (54)',
       'PUBLIC ADMINISTRATION (92)', 'REAL ESTATE, RENTAL & LEASING (53)',
       'RETAIL-BUILD. MATERIAL, GARDEN EQUIP. & SUPPLIES DEALERS (444)',
       'RETAIL-CLOTHING & CLOTHING ACCESSORIES STORES (448)',
       'RETAIL-ELECTRONICS & APPLIANCE STORES (443)',
       

In [7]:
# fix column names
#new_names = [replace(item) for item in list(sectors_q4_2020.columns)]
new_names = ['L_ACCOMMODATION_20', 'S_ADMIN_SUPPORT_20','A_AG_WILDLIFE_20','L_CULTURAL_REC_20','C_CONSTRUCTION_20',
'E_EDUCATION_20','S_FINANCIAL_20','L_RSTRNT_BAR_20','H_HEALTH_CARE_20','S_IT_DATA_20','S_CORPORATE_MGMT_20',
'M_MANUFACTURING_20','E_EXTRACTION_20','O_OTHER_20','X_ADJUSTMENTS_20','R_AUTO_PRIVATE_20','S_PROF_TECH_SRV_20',
'G_GOVERNMENT_20','S_RENTAL_LEASING_20','R_BUILDING_SUPPLY_20','R_CLOTHING_20','R_ELECTRONICS_20','R_GROCERY_BEV_20',
'R_FURNITURE_20','R_GAS_STATIONS_20','R_GENERAL_RETAIL_20','R_HEALTH_RETAIL_20','R_OTHER_RETAIL_20','R_AUTO_RETAIL_20',
'R_NONSTORE_RETAIL_20','R_SPORT_HOBBY_20','L_SPECIAL_EVENT_20','W_DISTRIBUTION_20','X_UNKNOWN_20','U_UTILITIES_20',
'W_WHLSALE_DURABLE_20','S_WHLSLE_ETRADE_20','W_WHLSLE_NDURABLE_20']
sectors_calyr_2020.columns = new_names

# get the total sales tax revenue
sectors_calyr_2020['TOTAL_20'] = sectors_calyr_2020[new_names].sum(axis=1)

sectors_calyr_2020.head()

L_ACCOMMODATION_20  S_ADMIN_SUPPORT_20  A_AG_WILDLIFE_20  \
City                                                                      
American Fork           1878400.0           2839307.0         2875138.0   
Beaver City             4912871.0             82679.0          500000.0   
Blanding                2075502.0            168578.0            1000.0   
Bountiful                700000.0           6271951.0           25681.0   
Brigham                 3657247.0            765247.0            8598.0   

               L_CULTURAL_REC_20  C_CONSTRUCTION_20  E_EDUCATION_20  \
City                                                                  
American Fork          5931456.0         39603098.0        442370.0   
Beaver City              30000.0           258778.0         46482.0   
Blanding                150000.0           878937.0        322791.0   
Bountiful              3904858.0          7206423.0        455012.0   
Brigham                1838897.0          3715796.0        204223.0   

               S_FINANCIAL_20  L_RSTRNT_BAR_20  H_HEALTH_CARE_20  \
City                                                               
American Fork      10592193.0      104349884.0         1898850.0   
Beaver City         1714929.0       10127155.0          250000.0   
Blanding              76907.0        1710742.0          200000.0   
Bountiful           4710201.0       50325246.0         4957256.0   
Brigham             1131414.0       27683037.0         1427040.0   

               S_IT_DATA_20  S_CORPORATE_MGMT_20  M_MANUFACTURING_20  \
City                                                                   
American Fork    31899515.0              60000.0          14442382.0   
Beaver City       2876142.0               8000.0           6259592.0   
Blanding          1815107.0               8000.0            855953.0   
Bountiful        23983623.0              90000.0          10655385.0   
Brigham          11221430.0              30000.0          30249675.0   

               E_EXTRACTION_20  O_OTHER_20  X_ADJUSTMENTS_20  \
City                                                           
American Fork         350000.0  47243005.0         9767823.0   
Beaver City             3000.0   3905297.0          840675.0   
Blanding                1000.0   1787098.0         -345158.0   
Bountiful              90000.0  25398674.0         8551645.0   
Brigham                15000.0  14024448.0         3115698.0   

               R_AUTO_PRIVATE_20  S_PROF_TECH_SRV_20  G_GOVERNMENT_20  \
City                                                                    
American Fork         18000000.0          20816225.0         200000.0   
Beaver City            2000000.0            586751.0              0.0   
Blanding               3000000.0            545856.0        2000000.0   
Bountiful             18250000.0          15004807.0        2000000.0   
Brigham               10250000.0           3180936.0           6000.0   

               S_RENTAL_LEASING_20  R_BUILDING_SUPPLY_20  R_CLOTHING_20  \
City                                                                      
American Fork           15322550.0            94422620.0     28914533.0   
Beaver City               582360.0             3335406.0       473077.0   
Blanding                  246818.0             8863939.0       362055.0   
Bountiful                8367435.0             7433117.0     12605951.0   
Brigham                  7562648.0             9629247.0      2060819.0   

               R_ELECTRONICS_20  R_GROCERY_BEV_20  R_FURNITURE_20  \
City                                                                
American Fork        26609471.0        17067144.0      16732105.0   
Beaver City            128134.0         7568481.0         49329.0   
Blanding               442393.0         8000000.0         57899.0   
Bountiful            16355255.0        99440715.0       7156565.0   
Brigham               5633194.0        48161414.0       2217817.0   

               R_GAS_STATIONS_20  R_GE

#### Load Q4 2019 sales tax revenue data

In [8]:
# load excel sheet into a dataframe
xlsx = '.\\Inputs\\2019-annual-sales.xlsx'
calyr_2019 = pd.read_excel(xlsx, sheet_name='Table 9', header=5)

# Drop last empty two rows
calyr_2019.drop(calyr_2019.tail(2).index,inplace=True)

# forward fill values from merged cells
calyr_2019['County'] = calyr_2019['County'].fillna(method='ffill')
calyr_2019['Location Code'] = calyr_2019['Location Code'].fillna(method='ffill')
calyr_2019['City'] = calyr_2019['City'].fillna(method='ffill')

# Figures with less than 10 taxpayers have been rounded up per Tax Commission disclosure rules.
# Rename column and convert values to boolean
calyr_2019.rename(columns={'Unnamed: 5':'Rounded Up'}, inplace=True)
crosswalk = {'*':True, np.nan:False} # County_ID
calyr_2019['Rounded Up'] = calyr_2019['Rounded Up'].map(crosswalk)

calyr_2019.tail()

County  Location Code         City  \
2445  Weber County        29040.0  South Ogden   
2446  Weber County        29040.0  South Ogden   
2447  Weber County        29040.0  South Ogden   
2448  Weber County        29040.0  South Ogden   
2449  Weber County        29040.0  South Ogden   

                          Economic Sector (NAICS Code)     CY2019  Rounded Up  
2445  OTHER SERVICES-EXCEPT PUBLIC ADMINISTRATION (81)  5865210.0       False  
2446                        PUBLIC ADMINISTRATION (92)     7000.0        True  
2447                       PRIVATE MOTOR VEHICLE SALES  5000000.0        True  
2448                               SPECIAL EVENT SALES    80000.0        True  
2449                   PRIOR-PERIOD PAYMENTS & REFUNDS  1648584.0       False

In [9]:
# pivot the table to get sectors as columns with tax values by city
sectors_calyr_2019 = pd.pivot_table(calyr_2019,values = 'CY2019',index ='City', columns = 'Economic Sector (NAICS Code)', 
                                 aggfunc='first')

# fill NAs with 0
sectors_calyr_2019 = sectors_calyr_2019.fillna(0)

# fix column names
#new_names = [replace(item) for item in list(sectors_q4_2019.columns)]
new_names = ['L_ACCOMMODATION_19', 'S_ADMIN_SUPPORT_19','A_AG_WILDLIFE_19','L_CULTURAL_REC_19','C_CONSTRUCTION_19',
'E_EDUCATION_19','S_FINANCIAL_19','L_RSTRNT_BAR_19','H_HEALTH_CARE_19','S_IT_DATA_19','S_CORPORATE_MGMT_19',
'M_MANUFACTURING_19','E_EXTRACTION_19','O_OTHER_19','X_ADJUSTMENTS_19','R_AUTO_PRIVATE_19','S_PROF_TECH_SRV_19',
'G_GOVERNMENT_19','S_RENTAL_LEASING_19','R_BUILDING_SUPPLY_19','R_CLOTHING_19','R_ELECTRONICS_19','R_GROCERY_BEV_19',
'R_FURNITURE_19','R_GAS_STATIONS_19','R_GENERAL_RETAIL_19','R_HEALTH_RETAIL_19','R_OTHER_RETAIL_19','R_AUTO_RETAIL_19',
'R_NONSTORE_RETAIL_19','R_SPORT_HOBBY_19','L_SPECIAL_EVENT_19','W_DISTRIBUTION_19','X_UNKNOWN_19','U_UTILITIES_19',
'W_WHLSALE_DURABLE_19','S_WHLSLE_ETRADE_19','W_WHLSLE_NDURABLE_19']
sectors_calyr_2019.columns = new_names

# drop unknown column since its not present in 2020
#sectors_q4_2019.drop('X_UNKNOWN_19', axis=1, inplace=True)
#new_names.remove('X_UNKNOWN_19')

# get the total sales tax revenue
sectors_calyr_2019['TOTAL_19'] = sectors_calyr_2019[new_names].sum(axis=1)

sectors_calyr_2019.head()

L_ACCOMMODATION_19  S_ADMIN_SUPPORT_19  A_AG_WILDLIFE_19  \
City                                                                      
American Fork           2537613.0           2552910.0         1750000.0   
Beaver City             4805537.0            103632.0          300000.0   
Blanding                4043564.0            204922.0            2000.0   
Bountiful                700000.0           9536828.0            6000.0   
Brigham                 5887485.0            763477.0            5000.0   

               L_CULTURAL_REC_19  C_CONSTRUCTION_19  E_EDUCATION_19  \
City                                                                  
American Fork          6467601.0         32735467.0        416107.0   
Beaver City              30000.0           360477.0         68551.0   
Blanding                250000.0           815000.0        467154.0   
Bountiful              4409965.0          7437407.0        455409.0   
Brigham                1993021.0          5949405.0        239569.0   

               S_FINANCIAL_19  L_RSTRNT_BAR_19  H_HEALTH_CARE_19  \
City                                                               
American Fork      10128930.0       97383158.0         1918417.0   
Beaver City         1190849.0        9820732.0          200000.0   
Blanding              86263.0        2250000.0          250000.0   
Bountiful           4872026.0       52253762.0         4107293.0   
Brigham             1145653.0       26906567.0         1666410.0   

               S_IT_DATA_19  S_CORPORATE_MGMT_19  M_MANUFACTURING_19  \
City                                                                   
American Fork    38370756.0              60000.0          10741593.0   
Beaver City       2380564.0               6000.0           4461526.0   
Blanding          2174354.0              15000.0            609232.0   
Bountiful        20506638.0             100000.0          10940271.0   
Brigham          11071258.0              70000.0          26059670.0   

               E_EXTRACTION_19  O_OTHER_19  X_ADJUSTMENTS_19  \
City                                                           
American Fork         250000.0  43377648.0        11906936.0   
Beaver City                0.0   3456681.0          628826.0   
Blanding                   0.0   1079534.0         1697521.0   
Bountiful              70000.0  26445564.0        11073265.0   
Brigham                10000.0  12630258.0         4451055.0   

               R_AUTO_PRIVATE_19  S_PROF_TECH_SRV_19  G_GOVERNMENT_19  \
City                                                                    
American Fork         11250000.0          17729496.0          90000.0   
Beaver City            1750000.0            400435.0              0.0   
Blanding               3000000.0            366851.0        2000000.0   
Bountiful             17500000.0          10914337.0        1500000.0   
Brigham                8250000.0           2059740.0           8000.0   

               S_RENTAL_LEASING_19  R_BUILDING_SUPPLY_19  R_CLOTHING_19  \
City                                                                      
American Fork           15136893.0            77332237.0     30782170.0   
Beaver City               399192.0             2207222.0       309900.0   
Blanding                  363550.0             6598780.0       264690.0   
Bountiful                9182908.0             4746627.0     12359267.0   
Brigham                  5932594.0             7006638.0      1857215.0   

               R_ELECTRONICS_19  R_GROCERY_BEV_19  R_FURNITURE_19  \
City                                                                
American Fork        32879118.0        16177423.0      15578247.0   
Beaver City            104804.0         6500000.0         25928.0   
Blanding               312800.0         7000000.0         35319.0   
Bountiful            14292963.0        89495871.0       6756247.0   
Brigham               4844330.0        41907498.0       2058474.0   

               R_GAS_STATIONS_19  R_GE

### Calculate the differences 
*2020 - 2019 = Diff*

In [10]:
# Check join output shape
print(sectors_calyr_2019.shape)
print(sectors_calyr_2020.shape)

(66, 39)
(66, 39)


In [11]:
# merge the 2019-2020 difference with the sdf
sectors_calyr_diff = pd.DataFrame(sectors_calyr_2020.values - sectors_calyr_2019.values)

# add back the column names
new_names = ['L_ACCOMMODATION_D20', 'S_ADMIN_SUPPORT_D20','A_AG_WILDLIFE_D20','L_CULTURAL_REC_D20','C_CONSTRUCTION_D20',
'E_EDUCATION_D20','S_FINANCIAL_D20','L_RSTRNT_BAR_D20','H_HEALTH_CARE_D20','S_IT_DATA_D20','S_CORPORATE_MGMT_D20',
'M_MANUFACTURING_D20','E_EXTRACTION_D20','O_OTHER_D20','X_ADJUSTMENTS_D20','R_AUTO_PRIVATE_D20','S_PROF_TECH_SRV_D20',
'G_GOVERNMENT_D20','S_RENTAL_LEASING_D20','R_BUILDING_SUPPLY_D20','R_CLOTHING_D20','R_ELECTRONICS_D20','R_GROCERY_BEV_D20',
'R_FURNITURE_D20','R_GAS_STATIONS_D20','R_GENERAL_RETAIL_D20','R_HEALTH_RETAIL_D20','R_OTHER_RETAIL_D20','R_AUTO_RETAIL_D20',
'R_NONSTORE_RETAIL_D20','R_SPORT_HOBBY_D20','L_SPECIAL_EVENT_D20','W_DISTRIBUTION_D20','X_UNKNOWN_D20','U_UTILITIES_D20',
'W_WHLSALE_DURABLE_D20','S_WHLSLE_ETRADE_D20','W_WHLSLE_NDURABLE_D20', 'TOTAL_D20']

sectors_calyr_diff.columns = new_names

# add back cities as index
sectors_calyr_diff['City'] = sectors_calyr_2020.index
sectors_calyr_diff = sectors_calyr_diff.set_index('City')

#check table
sectors_calyr_diff.head(10)

L_ACCOMMODATION_D20  S_ADMIN_SUPPORT_D20  \
City                                                           
American Fork                 -659213.0             286397.0   
Beaver City                    107334.0             -20953.0   
Blanding                     -1968062.0             -36344.0   
Bountiful                           0.0           -3264877.0   
Brigham                      -2230238.0               1770.0   
Cedar City                   -4223407.0            2000550.0   
Centerville                     20000.0             236928.0   
Clearfield                    -235020.0             -30012.0   
Cottonwood Heights            -183999.0             259164.0   
Delta                               0.0             -11691.0   

                    A_AG_WILDLIFE_D20  L_CULTURAL_REC_D20  C_CONSTRUCTION_D20  \
City                                                                            
American Fork               1125138.0           -536145.0           6867631.0   
Beaver City                  200000.0                 0.0           -101699.0   
Blanding                      -1000.0           -100000.0             63937.0   
Bountiful                     19681.0           -505107.0           -230984.0   
Brigham                        3598.0           -154124.0          -2233609.0   
Cedar City                     7625.0            -92623.0           1798180.0   
Centerville                   -2253.0             53889.0           -928230.0   
Clearfield                     6393.0            327405.0           1317343.0   
Cottonwood Heights             7711.0          -3376888.0            535984.0   
Delta                        100000.0            -50000.0           1773999.0   

                    E_EDUCATION_D20  S_FINANCIAL_D20  L_RSTRNT_BAR_D20  \
City                                                                     
American Fork               26263.0         463263.0         6966726.0   
Beaver City                -22069.0         524080.0          306423.0   
Blanding                  -144363.0          -9356.0         -539258.0   
Bountiful                    -397.0        -161825.0        -1928516.0   
Brigham                    -35346.0         -14239.0          776470.0   
Cedar City                -905809.0        -764229.0         3980672.0   
Centerville                 48168.0        -636127.0          828467.0   
Clearfield                  24350.0         -22082.0          584800.0   
Cottonwood Heights         -43283.0        -472220.0       -10073738.0   
Delta                        2000.0         -40366.0          389717.0   

                    H_HEALTH_CARE_D20  S_IT_DATA_D20  S_CORPORATE_MGMT_D20  \
City                                                                         
American Fork                -19567.0     -6471241.0                   0.0   
Beaver City                   50000.0       495578.0                2000.0   
Blanding                     -50000.0      -359247.0               -7000.0   
Bountiful                    849963.0      3476985.0              -10000.0   
Brigham                     -239370.0       150172.0              -40000.0   
Cedar City                   114207.0       878024.0              -10000.0   
Centerville                  -40606.0     -5914632.0             -380000.0   
Clearfield                    -1948.0      2824892.0              750000.0   
Cottonwood Heights          -278704.0      3946154.0             -250000.0   
Delta                             0.0        98344.0              -40000.0   

                    M_MANUFACTURING_D20  E_EXTRACTION_D20  O_OTHER_D20  \
City                                                                     
American Fork                 3700789.0          100000.0    3865357.0   
Beaver City                   1798066.0            3000.0     448616.0   
Blanding                       246721.0            1000.0     707564.0   
Bountiful                     -284886.0           20000.0   -1046890.0   
Brigham 

### Load municipalities and townships shapefile and join to formatted sales tax data

In [13]:
# load cities shapefile into pandas spatial dataframe
cities_shp = '.\\Inputs\\Cities.shp'
cities_sdf = pd.DataFrame.spatial.from_featureclass(cities_shp)

# Rename columns
cities_sdf.columns = ['FID', 'NAME', 'SOURCE', 'SALESTAXID', 'POPLASTCEN', 'POPLASTEST',
       'AREA_SQMI', 'SHAPE']

In [14]:
# merge the 2020 sdf with the pivoted data
cities_sdf2 = cities_sdf.merge(sectors_calyr_2019, left_on= "NAME", right_on="City", how="inner")
cities_sdf2 = cities_sdf2.merge(sectors_calyr_2020, left_on= "NAME", right_on="City", how="inner")
cities_sdf2 = cities_sdf2.merge(sectors_calyr_diff, left_on= "NAME", right_on="City", how="inner")
cities_sdf2.tail(10)


FID              NAME  SOURCE SALESTAXID  POPLASTCEN  POPLASTEST  \
55  222          Syracuse  Cities        049       25113       30400   
56  224      Taylorsville  Cities        142       59989       60192   
57  225       Tooele City  Cities        048       32629       35251   
58  228         Tremonton  Cities        113        7828        8882   
59  232            Vernal  Cities        024        9460       10370   
60  238   Washington City  Cities        027       19640       27686   
61  243    West Bountiful  Cities        061        5420        5731   
62  245       West Jordan  Cities        155      105891      116046   
63  247  West Valley City  Cities        167      133796      136401   
64  252       Woods Cross  Cities        057       10086       11328   

    AREA_SQMI                                              SHAPE  \
55  10.193900  {'rings': [[[411406.60675760085, 4549994.30436...   
56  10.839700  {'rings': [[[422191.826826254, 4504303.3642982...   
57  24.087099  {'rings': [[[392501.67668449413, 4486111.93423...   
58   7.955100  {'rings': [[[399303.9767575987, 4621607.124490...   
59   4.632710  {'rings': [[[627188.8071560834, 4479658.624106...   
60  36.165600  {'rings': [[[278123.416362615, 4102773.6636791...   
61   3.299440  {'rings': [[[421904.10680578306, 4528452.80425...   
62  32.306999  {'rings': [[[415038.796736156, 4499131.1642154...   
63  35.799702  {'rings': [[[418708.9467166035, 4508796.864202...   
64   3.835550  {'rings': [[[424779.70679974474, 4524289.10421...   

    L_ACCOMMODATION_19  S_ADMIN_SUPPORT_19  A_AG_WILDLIFE_19  \
55            150000.0            317480.0            4000.0   
56           1500000.0           1935311.0           40000.0   
57           4822863.0            901317.0           40000.0   
58           3547899.0           4197828.0           25000.0   
59           9247151.0            684128.0           15000.0   
60           9295636.0            758242.0          800000.0   
61           2500000.0             86459.0          150000.0   
62           6972329.0           4962893.0         1216151.0   
63          30178782.0          13267984.0          150000.0   
64           5248865.0           1428244.0            2000.0   

    L_CULTURAL_REC_19  C_CONSTRUCTION_19  E_EDUCATION_19  S_FINANCIAL_19  \
55          3848890.0           814638.0         92845.0       1879781.0   
56          6954210.0          5321911.0        488058.0       8850063.0   
57          4511291.0          4087933.0        156135.0       3473180.0   
58          1750000.0          3664774.0         30000.0        512930.0   
59          1033612.0          1101023.0         99980.0       1546401.0   
60          3694843.0          3880933.0         23567.0       1744891.0   
61          1000000.0           633546.0         20559.0        444268.0   
62          9414486.0         40915630.0        605898.0       9872931.0   
63         16551341.0         99919145.0       1154734.0      14683353.0   
64             1000.0          3932768.0        281098.0        959328.0   

    L_RSTRNT_BAR_19  H_HEALTH_CARE_19  S_IT_DATA_19  S_CORPORATE_MGMT_19  \
55       18287033.0          152701.0    11396365.0              70000.0   
56      111116485.0          522004.0    36946094.0              90000.0   
57       48737530.0         3290064.0    18788836.0              60000.0   
58       12363852.0           15000.0     4032957.0              15000.0   
59       40297076.0         1250143.0    19244286.0              30000.0   
60       36567045.0           38715.0    11350935.0              70000.0   
61       15639883.0           25000.0     6639179.0              25000.0   
62      148199045.0         4210306.0    88225999.0             900000.0   
63      223789508.0         5603292.0   105103387.0             150000.0   
64       19274620.0          150000.0     9970173.0              10000.0   

    M_MANUFACTURING_19  E_EXTRACTION_19  O_OTHER_19  X_ADJUSTMENTS_19  \
55        

In [15]:
# Confirm join count, Salt Lake County (Unincorporated) should be the only one not joined
l1 = list(calyr_2020['City'].value_counts().index)
l2 = list(cities_sdf['NAME'].value_counts().index)
list(set(l1) - set(l2))

['Salt Lake County (Unincorporated)']

In [16]:
# Create file gdb and export sdf to feature class
outputs = '.\\Outputs'

gdb = os.path.join(outputs, "taxable_sales.gdb")
if not arcpy.Exists(gdb):
    arcpy.CreateFileGDB_management(outputs, "taxable_sales.gdb")

cities_sdf2.spatial.to_featureclass(location=os.path.join(outputs, "taxable_sales.gdb","taxable_sales_utah_calyr"))

'C:\\LocalGIS\\GitProjects\\UT_Sales_Tax_Analysis\\Outputs\\taxable_sales.gdb\\taxable_sales_utah_calyr'

### Analysis

In [17]:
cities_sdf2

FID              NAME  SOURCE SALESTAXID  POPLASTCEN  POPLASTEST  \
0     5     American Fork  Cities        002       27413       32519   
1    12       Beaver City  Cities        002        3137        3104   
2    15          Blanding  Cities        002        3514        3696   
3    19         Bountiful  Cities        004       43480       44098   
4    21           Brigham  Cities        017       18277       19404   
..  ...               ...     ...        ...         ...         ...   
60  238   Washington City  Cities        027       19640       27686   
61  243    West Bountiful  Cities        061        5420        5731   
62  245       West Jordan  Cities        155      105891      116046   
63  247  West Valley City  Cities        167      133796      136401   
64  252       Woods Cross  Cities        057       10086       11328   

    AREA_SQMI                                              SHAPE  \
0   11.256600  {"rings": [[[434244.49682960496, 4467294.37416...   
1    6.638160  {"rings": [[[355522.91657235345, 4239879.43388...   
2   13.205500  {"rings": [[[634403.6869837723, 4163230.893496...   
3   13.209000  {"rings": [[[426520.60679261835, 4528697.30425...   
4   24.910801  {"rings": [[[415000.02685977484, 4598737.23444...   
..        ...                                                ...   
60  36.165600  {"rings": [[[278123.416362615, 4102773.6636791...   
61   3.299440  {"rings": [[[421904.10680578306, 4528452.80425...   
62  32.306999  {"rings": [[[415038.796736156, 4499131.1642154...   
63  35.799702  {"rings": [[[418708.9467166035, 4508796.864202...   
64   3.835550  {"rings": [[[424779.70679974474, 4524289.10421...   

    L_ACCOMMODATION_19  S_ADMIN_SUPPORT_19  A_AG_WILDLIFE_19  \
0            2537613.0           2552910.0         1750000.0   
1            4805537.0            103632.0          300000.0   
2            4043564.0            204922.0            2000.0   
3             700000.0           9536828.0            6000.0   
4            5887485.0            763477.0            5000.0   
..                 ...                 ...               ...   
60           9295636.0            758242.0          800000.0   
61           2500000.0             86459.0          150000.0   
62           6972329.0           4962893.0         1216151.0   
63          30178782.0          13267984.0          150000.0   
64           5248865.0           1428244.0            2000.0   

    L_CULTURAL_REC_19  C_CONSTRUCTION_19  E_EDUCATION_19  S_FINANCIAL_19  \
0           6467601.0         32735467.0        416107.0      10128930.0   
1             30000.0           360477.0         68551.0       1190849.0   
2            250000.0           815000.0        467154.0         86263.0   
3           4409965.0          7437407.0        455409.0       4872026.0   
4           1993021.0          5949405.0        239569.0       1145653.0   
..                ...                ...             ...             ...   
60          3694843.0          3880933.0         23567.0       1744891.0   
61          1000000.0           633546.0         20559.0        444268.0   
62          9414486.0         40915630.0        605898.0       9872931.0   
63         16551341.0         99919145.0       1154734.0      14683353.0   
64             1000.0          3932768.0        281098.0        959328.0   

    L_RSTRNT_BAR_19  H_HEALTH_CARE_19  S_IT_DATA_19  S_CORPORATE_MGMT_19  \
0        97383158.0         1918417.0    38370756.0              60000.0   
1         9820732.0          200000.0     2380564.0               6000.0   
2         2250000.0          250000.0     2174354.0              15000.0   
3        52253762.0         4107293.0    20506638.0             100000.0   
4        26906567.0         1666410.0    11071258.0              70000.0   
..              ...               ...           ...                  ...   
60       36567045.0           38715.0    11350935.0              70000.0   
61       15639883.0           25

In [18]:
# Get year specific column names
standard_cols = ['FID','NAME','SOURCE','SALESTAXID','POPLASTCEN','POPLASTEST','AREA_SQMI', 'SHAPE']
names_19 = standard_cols + [col for col in cities_sdf2.columns if '_19' in col]
names_20 = standard_cols + [col for col in cities_sdf2.columns if '_20' in col]
names_D20 = standard_cols + [col for col in cities_sdf2.columns if '_D20' in col]

# subset columns by time
cities_calyr_19 = cities_sdf2[names_19]
cities_calyr_20 = cities_sdf2[names_20]
cities_calyr_d20 = cities_sdf2[names_D20]


In [19]:
# 10 cities with smallest difference of revenue
cities_calyr_d20[['NAME', 'POPLASTCEN', 'AREA_SQMI', 'TOTAL_D20']].sort_values('TOTAL_D20', ascending=False).head(10)

NAME  POPLASTCEN  AREA_SQMI    TOTAL_D20
54         St George       76548  77.209900  430010075.0
62       West Jordan      105891  32.306999  299402288.0
20            Layton       69457  22.576200  201867725.0
50   South Salt Lake       24282   6.936550  189977146.0
34              Orem       92993  18.628000  187973628.0
51      Spanish Fork       36232  16.249701  187486973.0
23             Logan       49612  18.485800  186064535.0
63  West Valley City      133796  35.799702  183230415.0
21              Lehi       49315  28.898199  164776229.0
0      American Fork       27413  11.256600  162046226.0

In [20]:
# 10 cities with largest difference of revenue
cities_calyr_d20[['NAME', 'POPLASTCEN', 'AREA_SQMI', 'TOTAL_D20']].sort_values('TOTAL_D20', ascending=True).head(10)

NAME  POPLASTCEN   AREA_SQMI    TOTAL_D20
45   Salt Lake City      190749  110.734001 -311328076.0
35        Park City        7709   20.341700  -87245688.0
32  North Salt Lake       16930    8.512150  -49944002.0
30           Murray       47311   12.293700  -46516692.0
12       Farmington       18752   10.035100  -31079195.0
43        Roosevelt        6260    6.757070  -27292857.0
27             Moab        5124    4.802460  -13108412.0
52       Springdale         569    4.594880   -8203389.0
59           Vernal        9460    4.632710   -2417569.0
17            Kanab        4413   14.473200    -753309.0

#### Some industries of interest:
- RETAIL-GASOLINE STATIONS (447)
- RETAIL-FOOD & BEVERAGE STORES (445)
- REAL ESTATE, RENTAL & LEASING (53)
- ARTS, ENTERTAINMENT AND RECREATION (71)
- ACCOMMODATION (721)